### PR TITLE
convert GLES1.00 to GLES3.00

### LINKED ISSUE #1816: BlazorGL 'Shader Compilation Failed' runtime exception using fwidth function
Shaders that use the `fwidth` function build successfully with the content pipeline but fail at runtime in KNI BlazorGL with the exception 'Shader Compilation Failed'.

My best guess is the following is required to be added to the BlazorGL shader as it is in DesktopGL:

```
#extension GL_OES_standard_derivatives : enable
```

Here's an example shader that works on MonoGame DesktopGL, iOS, and Android but fails on KNI BlazorGL due to the `fwidth` function call.

```
#if OPENGL
	#define SV_POSITION POSITION
	#define VS_SHADERMODEL vs_3_0
	#define PS_SHADERMODEL ps_3_0
#else
	#define VS_SHADERMODEL vs_4_0_level_9_3
	#define PS_SHADERMODEL ps_4_0_level_9_3
#endif

float2 TextureSize;
float PxRange;
float Thickness;

sampler SpriteTextureSampler : register(s0);

struct VertexShaderOutput
{
	float4 Position : SV_POSITION;
	float4 Color : COLOR0;
	float2 TexCoord : TEXCOORD0;
};

float Median(float a, float b, float c)
{	
	return max(min(a, b), min(max(a, b), c));
}

float LinearStep(float a, float b, float x) {
	return clamp((x - a) / (b - a), 0.0, 1.0);
}

float4 SamplePS(VertexShaderOutput input) : COLOR
{
	float2 p = input.TexCoord;
	float pxSize = min(0.5 / PxRange * (fwidth(p.x) * TextureSize.x + fwidth(p.y) * TextureSize.y), 0.25);

	float3 samp = tex2D(SpriteTextureSampler, p).rgb;

	float sigDist = 2.0f * Median(samp.r, samp.g, samp.b) - 1.0f + Thickness;

	float opacity = LinearStep(-pxSize, pxSize, sigDist);

	return input.Color * opacity;
}

technique SmoothText
{
	pass P0
	{
		PixelShader = compile PS_SHADERMODEL SamplePS();		
	}
};
```

#### What version of KNI does the bug occur on:
- KNI 3.13.9001

#### What operating system are you using:
- Windows11

#### What KNI platform are you using:

- BlazorGL

### PR DESCRIPTION
**Pull request by nkast**
fixes #1816

### CODE REVIEW

**Comment by dtaylorus** [comment]
@nkast Does this fix the issue where you provide only a pixel shader and the vertex shader comes from SpriteEffect?

In that case, the vertex shader will be v1 and the pixel shader v3 which fails at runtime.

**Comment by nkast** [comment]
All shaders are patched at runtime, so it will work with any preexisted effect and library.

I'll leave this open for a while to get it tested and get any feedback.

File: Platforms/Graphics/.BlazorGL/Shader/ConcreteShader.cs [hunk 1]
```diff
@@ -51,6 +59,44 @@ internal void CreateShader(GraphicsContextStrategy contextStrategy, WebGLShaderT
...
         }
 
+        static Regex rgxOES = new Regex(
+                @"#extension GL_OES_standard_derivatives : enable", System.Text.RegularExpressions.RegexOptions.Multiline);
+        static Regex rgxAttribute = new Regex(
+            @"^attribute(?=\s)", System.Text.RegularExpressions.RegexOptions.Multiline);
+        static Regex rgxVarying = new Regex(
+            @"^varying(?=\s)", System.Text.RegularExpressions.RegexOptions.Multiline);
+        static Regex rgxFragColor = new Regex(
+            @"^#define ps_oC0 gl_FragColor", System.Text.RegularExpressions.RegexOptions.Multiline);
```
Comment (dtaylorus) [inline Platforms/Graphics/.BlazorGL/Shader/ConcreteShader.cs:69]: I'm not 100% sure that the variable name will always be `ps_oC0`. Maybe this variable name should be extracted for the replacement?

Comment (nkast) [reply Platforms/Graphics/.BlazorGL/Shader/ConcreteShader.cs]: good point. Maybe it's different when attaching multiple rendertargets.

**Comment by dtaylorus** [comment]
Awesome. That's very elegant to patch at runtime rather than build time so I assume it will work with the default shaders like SpriteEffect. Looking forward to trying it out in my project and removing my hack to duplicate the SpriteEffect source code.

**Review by dtaylorus — APPROVED** [review]

**Pull request merged**